### PR TITLE
DP-1171 - Add DPF example: Create a Custom PipelineAction

### PR DESCRIPTION
**Pull request by gvjoshi25**
DP-1172 - Add DPF example: Create a Custom PipelineInput
DP-1173 - Add DPF example: Create a Custom PipelineOutput
Update version in build.gradle

### CODE REVIEW

File: src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            .setInput(pipelineInput)
+            .setOutput(pipelineOutput)
+            .addAction(new UpcaseFieldsAction().addField("Country"));
+        
+        pipeline.run();
+    }
+    
+}
+
+class UpcaseFieldsAction extends PipelineAction {
```
Comment (DeleTaylor) [inline src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java]: Please make this class a static inner class of CustomPipelineAction.

File: src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    @Override
+    public Record toRecord() {
+        return null;
+    }
+
+}
+
+class UpcaseField extends FieldTransformer {
```
Comment (DeleTaylor) [inline src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java]: No need for this new class, we should be able to use `BasicFieldTransformer.upperCase()`.

File: src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    @Override
+    public void fromRecord(Record source) {
+    }
+
+    @Override
+    public Record toRecord() {
+        return null;
+    }
```
Comment (DeleTaylor) [inline src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java]: Let's implement to/fromRecord to show how to participate in JSON serialization.

File: src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
```
Comment (DeleTaylor) [inline src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java]: Let's also implement `public void generateJavaCode(JavaCodeBuilder code)` to show how to participate in optional code generation for those that need it.

**Review by DeleTaylor — CHANGES_REQUESTED** [review]

File: src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java [hunk 5]
```diff
@@ -32,56 +40,75 @@ public static void main(String[] args) {
...
+        }
 
-    @Override
-    public void fromRecord(Record source) {
-    }
+        @Override
+        public Record toRecord() {
+            Record record = new Record();
+
+            if (Util.isNotEmpty(fieldNameList)) {
```
Comment (DeleTaylor) [inline src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java]: Let's reduce the code by emitting an empty array if no fields are in `fieldNameList`.

File: src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java [hunk 6]
```diff
@@ -32,56 +40,75 @@ public static void main(String[] args) {
...
-}
+    static class UpcaseFieldsAction extends PipelineAction {
 
-class UpcaseFieldsAction extends PipelineAction {
+        private final List<String> fieldNameList;
 
-    private final List<String> fieldNameList;
+        protected UpcaseFieldsAction() {
+            super("transform-upcase-fields", "Upper Case Fields");
+            fieldNameList = new ArrayList<>();
```
Comment (DeleTaylor) [inline src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java]: Move this initialization up as part of the declaration line.

File: src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java [hunk 7]
```diff
@@ -32,56 +40,75 @@ public static void main(String[] args) {
...
+            fieldNameList.add(fieldName);
+            return this;
+        }
 
-    public UpcaseFieldsAction addField(String fieldName) {
-        fieldNameList.add(fieldName);
-        return this;
-    }
+        @Override
+        public void fromRecord(Record source) {
```
Comment (DeleTaylor) [inline src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java]: Let's move the `toRecord()` method above this one to help readability.

**Review by DeleTaylor — CHANGES_REQUESTED** [review]

File: src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java (hunk 5 above)
Comment (gvjoshi25) [reply src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java]: @DeleTaylor you meant I should remove this condition, right?

File: src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java (hunk 1 above)
Comment (gvjoshi25) [reply src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java]: Done

File: src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java (hunk 2 above)
Comment (gvjoshi25) [reply src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java]: Done.

File: src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java (hunk 3 above)
Comment (gvjoshi25) [reply src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java]: Done.

File: src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java (hunk 4 above)
Comment (gvjoshi25) [reply src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java]: Done.

File: src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java (hunk 5 above)
Comment (gvjoshi25) [reply src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java]: I have removed this condition. If no field, then empty array will be added in record.

File: src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java (hunk 6 above)
Comment (gvjoshi25) [reply src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java]: Done.

File: src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java (hunk 7 above)
Comment (gvjoshi25) [reply src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java]: Done.

File: src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CreateCustomPipelineAction.java [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        Pipeline pipeline = new Pipeline()
+                .setInput(pipelineInput)
+                .setOutput(pipelineOutput)
+                .addAction(new UpcaseFieldsAction().addField("Country"));
+
+        pipeline.run();
+
+        System.out.println("---------------------------------------------------------------------------------------------------------");
+
+        System.out.println("Generated Code:" + System.lineSeparator());
```
Comment (DeleTaylor) [inline src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CreateCustomPipelineAction.java]: Can we drop the `System.lineSeparator()` call (here and below)?  Are they needed?

File: src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CreateCustomPipelineAction.java [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+        System.out.println(Util.formatJson(pipeline.toJsonString()));
+
+    }
+
+    public static class UpcaseFieldsAction extends PipelineAction {
+
+        private final List<String> fieldNameList = new ArrayList<>();
+
+        public UpcaseFieldsAction() {
+            super("transform-upcase-fields", "Upper Case Fields");
```
Comment (DeleTaylor) [inline src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CreateCustomPipelineAction.java]: "Uppercase" is one word.

**Review by DeleTaylor — CHANGES_REQUESTED** [review]

File: src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CreateCustomPipelineAction.java (hunk 8 above)
Comment (gvjoshi25) [reply src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CreateCustomPipelineAction.java]: Removed all `System.lineSeparator()`.

File: src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CreateCustomPipelineAction.java (hunk 9 above)
Comment (gvjoshi25) [reply src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CreateCustomPipelineAction.java]: Done.

**Review by DeleTaylor — APPROVED** [review]

**Pull request merged**